### PR TITLE
fix(orderBy): Throw exception when orderBy is given a non array-like object.

### LINKED ISSUE #11255: Using orderBy on an object should cause an error or warning
Over in https://github.com/angular/angular.js/issues/1286 it was decided that `orderBy` should not be modified to work when doing an ng-repeat over an object.

However, trying to use `orderBy` on an object doesn't give any clue as to what you're doing wrong; the ordering simply silently fails to take effect: http://jsfiddle.net/uo7kym1n/1/

I just got bitten by this behaviour and the idea that using `orderBy` on an object was wrong was certainly not the first idea that came to my mind. I suspect the 50 people who +1ed https://github.com/angular/angular.js/issues/1286 are likewise people who lost time trying to make sense of why their orderBy wasn't working without Angular offering them any clues.

Would it not be reasonable to have `orderBy` throw an error or warning when used on an object, instructing that it should not be so used?

### PR DESCRIPTION
**Pull request by nhodges**
Throws exception when orderBy is given a non array-like object to sort.

Closes #11255.

[Review on Reviewable](https://reviewable.io/reviews/angular/angular.js/11719)

### CODE REVIEW

File: src/ng/filter/orderBy.js [hunk 1]
```diff
@@ -155,7 +155,7 @@
 orderByFilter.$inject = ['$parse'];
 function orderByFilter($parse) {
   return function(array, sortPredicate, reverseOrder) {
-    if (!(isArrayLike(array))) return array;
+    if (!(isArrayLike(array))) throw minErr('filter')('notarray', 'Expected array but received: {0}', array);
```
Comment (petebacondarwin) [inline src/ng/filter/orderBy.js]: Sadly this error code is specific to the `filterFilter` so we have to provide a new one and associated documentation in `docs/content/error` or alternatively update the documentation to make this error code more generic.

**Comment by frankweb** [comment]
@nhodges @petebacondarwin Hi guys, I've been working in the documentation needed for this pull request. Please look at this 

https://github.com/angular/angular.js/compare/master...frankweb:orderby-notarray-documentation

I'd like to contribute to this pull request by providing documentation.

**Comment by nhodges** [comment]
:+1: @frankweb for providing documentation, I updated my PR to be fast forwardable.

**Comment by frankweb** [comment]
@nhodges :+1: do you want to include my documentation in your PR?

**Comment by nhodges** [comment]
@frankweb You got it :)

**Comment by nhodges** [comment]
@frankweb looks like the task to build the docs is failing after merging, can you take a look?

**Comment by frankweb** [comment]
@nhodges Of course, I'll take a look in a while :)

File: src/ng/filter/orderBy.js [hunk 2]
```diff
@@ -177,7 +177,7 @@ orderByFilter.$inject = ['$parse'];
 function orderByFilter($parse) {
   return function(array, sortPredicate, reverseOrder) {
 
-    if (!(isArrayLike(array))) return array;
+    if (!(isArrayLike(array))) throw minErr('filter')('notarray', 'Expected array but received: {0}', array);
```
Comment (frankweb) [inline src/ng/filter/orderBy.js]: `grunt test` runs without errors if I change `minErr('filter')` to `minErr('orderBy')`

File: test/ng/filter/orderBySpec.js [hunk 3]
```diff
@@ -8,6 +8,11 @@ describe('Filter: orderBy', function() {
 
 
   describe('(Arrays)', function() {
+    it('should throw an exception if no array-like object is provided', function() {
+      expect(function() { orderBy({}); }).
+        toThrowMinErr('filter', 'notarray', 'Expected array but received: {}');
```
Comment (frankweb) [inline test/ng/filter/orderBySpec.js]: The same for this line, we need to change `'filter` to `orderBy`

File: docs/content/error/orderBy/notarray.ngdoc [hunk 4]
```diff
@@ -0,0 +1,53 @@
+@ngdoc error
+@name orderBy:notarray
+@fullName Not an array
+@description
+
+This error occurs when {@link ng.orderBy filter} is not used with an array:
```
Comment (gkalpak) [inline docs/content/error/orderBy/notarray.ngdoc]: _filter_  -->  _orderBy_

File: docs/content/error/orderBy/notarray.ngdoc [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+              },
+      two:   {
+              name: 'something 2'
+             },
+      three: {
+             name: 'something 3'
+             },
+    };
+
+    //if using underscore
```
Comment (gkalpak) [inline docs/content/error/orderBy/notarray.ngdoc]: I don't think we want to have an example using underscore (or any 3rd-party library).

Comment (petebacondarwin) [inline docs/content/error/orderBy/notarray.ngdoc]: I agree. Sorry

Comment (frankweb) [inline docs/content/error/orderBy/notarray.ngdoc]: Ok guys, no problem, do you have any other comment?

File: docs/content/error/orderBy/notarray.ngdoc [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    $scope.objToArray = = _.values(obj);
```
Comment (frankweb) [inline docs/content/error/orderBy/notarray.ngdoc]: We can remove the comment and change 

``` javascript
$scope.objToArray = _.values(obj);
```

to

``` javascript
$scope.objToArray = Object.keys(obj).map(function(a){
   return obj[a] ;
});
```

Doing this we don't need underscore. What do you think @petebacondarwin @gkalpak ?

**Comment by nhodges** [comment]
:+1: @frankweb thanks for the heads up. Fixed and we have a green build again.

**Comment by gkalpak** [comment]
@nhodges, could you take care of the unresolved comments ?

File: docs/content/error/orderBy/notarray.ngdoc [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    var obj = {
+      one:   {
+              name: 'something 1'
+              },
+      two:   {
+              name: 'something 2'
+             },
+      three: {
+             name: 'something 3'
+             },
```
Comment (gkalpak) [inline docs/content/error/orderBy/notarray.ngdoc]: Also, the indentation is funny for this object. TBH, I would change the example to something like:

```html
<!-- HTML -->
<label>
  Order by:
  <select ng-model="orderProp" ng-options="prop for prop in ['id', 'name']"></select>
</label>
<div ng-repeat="(key, value) in myObj | orderBy:orderProp">
  {{ key }} : {{ value }}
</div>
```

```js
/* JS */
angular.module('aModule', [])
  .controller('aController', function($scope) {
    var myObj = {
      one: {id: 1, name: 'Some thing'},
      two: {id: 2, name: 'Another thing'},
      three: {id: 3, name: 'A third thing'}
    };

    $scope.arrFromMyObj = Object.keys(myObj).map(function(key) {
      return myObj[key];
    });
  });
```

**Comment by nhodges** [comment]
@gkalpak You got it. Pushed.

**Comment by nhodges** [comment]
@gkalpak @petebacondarwin @frankweb 

Anything else I can take care of?

**Comment by frankweb** [comment]
@nhodges of course, I'll take care of this, do you mind if I do a new PR using your branch?

**Comment by frankweb** [comment]
@nhodges Sorry for the last comment, I already deleted it, my comment was about other PR.

**Comment by frankweb** [comment]
There is only one problem with the PR, I think all the commits should be made from one account, since you merged my branch this PR has commits with two accounts. Because of this, cla label says `no` @Narretz @petebacondarwin should @nhodges do another PR or what should he do?

**Comment by nhodges** [comment]
Hey Francisco,

I think if you just sign the CLA should be good?

Made with thumbs.

**Comment by frankweb** [comment]
@nhodges I guess we should comment something to the googlebot, but I don't know exactly what.

**Comment by frankweb** [comment]
@googlebot I'm okay with these commits.

**Comment by frankweb** [comment]
@googlebot I authored these commits

**Comment by nhodges** [comment]
@frankweb 

Hey Francisco, can you try clicking the link below and filling out the form?

http://code.google.com/legal/individual-cla-v1.0.html

Thanks!

**Comment by frankweb** [comment]
@nhodges Of course, but I'm already signed in. I'll try again.

**Comment by petebacondarwin** [comment]
@nhodges and @frankweb - as long as you have both signed the CLA then this is fine. The Google Bot always gets confused when a PR contains commits from more than one author.

**Comment by frankweb** [comment]
@petebacondarwin Thank you it's good to know that. Do you need another change to this PR?

File: docs/content/error/orderBy/notarray.ngdoc [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+This error occurs when {@link ng.orderBy orderBy} is not used with an array:
+```html
+<input ng-model="search">
+<div ng-repeat="(key, value) in myObj | orderBy:search">
+  {{ key }} : {{ value }}
+</div>
+```
+
+orderBy must be used with an array so a subset of items can be returned.
```
Comment (petebacondarwin) [inline docs/content/error/orderBy/notarray.ngdoc]: `orderBy must be used with an array-like input so a subset of items can be returned.`

File: docs/content/error/orderBy/notarray.ngdoc [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+The array can be initialized asynchronously and therefore null or undefined won't throw this error.
```
Comment (petebacondarwin) [inline docs/content/error/orderBy/notarray.ngdoc]: use backticks for `null` and `undefined`

File: docs/content/error/orderBy/notarray.ngdoc [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+To orderBy an object by the value of its properties you can create your own array based on that object:
```
Comment (petebacondarwin) [inline docs/content/error/orderBy/notarray.ngdoc]: To use `orderBy` to order the properties of an object, you can create...

File: docs/content/error/orderBy/notarray.ngdoc [hunk 11]
```diff
@@ -0,0 +1,50 @@
+@ngdoc error
+@name orderBy:notarray
+@fullName Not an array
```
Comment (petebacondarwin) [inline docs/content/error/orderBy/notarray.ngdoc]: Value is not array-like

File: docs/content/error/orderBy/notarray.ngdoc [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+@description
+
+This error occurs when {@link ng.orderBy orderBy} is not used with an array:
```
Comment (petebacondarwin) [inline docs/content/error/orderBy/notarray.ngdoc]: is not passed an array-like value

File: test/ng/filter/orderBySpec.js [hunk 13]
```diff
@@ -8,6 +8,11 @@ describe('Filter: orderBy', function() {
 
 
   describe('(Arrays)', function() {
+    it('should throw an exception if no array-like object is provided', function() {
+      expect(function() { orderBy({}); }).
+        toThrowMinErr('orderBy', 'notarray', 'Expected array but received: {}');
+    });
+
```
Comment (petebacondarwin) [inline test/ng/filter/orderBySpec.js]: Please add a test that shows that an array-like (i.e. not a pure array) does not fail, unless there is already such a test in the spec.

**Comment by petebacondarwin** [comment]
I have added a few minor comments.
Can you also update the documentation comment to make it clear that people can pass in array-like objects and that a non-array like object will result in an error?
Finally, since this is a breaking change, we need a BREAKING CHANGES block in one of the commits, if you could provide one.
We can squash all these commits together when we merge. It is likely to go into one commit. Can @nhodges and @frankweb agree on who would be the final author of this?

**Comment by nhodges** [comment]
@petebacondarwin @frankweb 

I went ahead and took care of the requested changes to docs and added a test that takes in a strong and returns a sorted  array.

**Comment by frankweb** [comment]
@nhodges Good! Please squash the commits, this is your pull request.

File: docs/content/error/orderBy/notarray.ngdoc [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+    });
+  });
+```
+That can be used as:
+```html
+<label>
+  Order by:
+  <select ng-model="orderProp" ng-options="prop for prop in ['id', 'name']"></select>
+</label>
+<div ng-repeat="(key, value) in myObj | orderBy:orderProp">
```
Comment (gkalpak) [inline docs/content/error/orderBy/notarray.ngdoc]: _myObj_  -->  _arrFromMyObj_

File: docs/content/error/orderBy/notarray.ngdoc [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+  {{ key }} : {{ value }}
+</div>
+```
+
+You could as well convert the object to an array using a filter such as
+[toArrayFilter](https://github.com/petebacondarwin/angular-toArrayFilter):
+```html
+<input ng-model="search">
+<div ng-repeat="item in myObj | toArray:false | filter:search">
```
Comment (gkalpak) [inline docs/content/error/orderBy/notarray.ngdoc]: _filter:search_  -->  _orderBy:search_

**Comment by gkalpak** [comment]
With https://github.com/angular/angular.js/pull/11719#issuecomment-150649515, I think we are good to go :smile:
(I left  couple of minor comments (typos), but I will take care of them while merging.)

Thx @nhodges and @frankweb (and @petebacondarwin) for working on this one !

**Comment by gkalpak** [comment]
@nhodges, actually could you rebase your changes on master. It's been a long time since the first commit...

**Comment by nhodges** [comment]
@gkalpak rebased. LMK if I can do anything else or if you want me to take care of the changes above.

File: src/ng/filter/orderBy.js [hunk 16]
```diff
@@ -177,7 +177,7 @@ orderByFilter.$inject = ['$parse'];
 function orderByFilter($parse) {
   return function(array, sortPredicate, reverseOrder) {
 
-    if (!(isArrayLike(array))) return array;
+    if (!(isArrayLike(array))) throw minErr('orderBy')('notarray', 'Expected array but received: {0}', array);
```
Comment (gkalpak) [inline src/ng/filter/orderBy.js]: I was in the middle of merging when I realized there is an issue here: We need to allow `null`/`undefined` (in order to support asnc loading). 
This block needs to be modified (see [filter#L132-L138][1] for an example).

[1]: https://github.com/angular/angular.js/blob/3ca4ca463c97a2bbca09f0366d2a753a13751143/src/ng/filter/filter.js#L132-L138

**Comment by gkalpak** [comment]
@nhodges, there seems to be one last little thing to do :smile:
Could you take care of https://github.com/angular/angular.js/pull/11719#discussion_r43314775 and add a test that verifies `null`/`undefined` do not result in an error ?

File: src/ng/filter/orderBy.js (hunk 16 above)
Comment (frankweb) [inline src/ng/filter/orderBy.js]: @gkalpak  I guess something like 

``` javascript
if(!array) return array;

if (!(isArrayLike(array))) throw minErr('orderBy')('notarray', 'Expected array but received: {0}', array);
```

would be easier to read than the example.

This is just an idea,of course :) what do you think about that?

Comment (petebacondarwin) [inline src/ng/filter/orderBy.js]: I think that is about right and similar to what we do in other filters.

Comment (gkalpak) [inline src/ng/filter/orderBy.js]: In `filterFilter` we only allow null, undefined and array-like (and I think that should we do here as well). So it would be more like:

```js
if (array == null) return array;
```

**Comment by nhodges** [comment]
@gkalpak Sure thing, please take a look at the latest changeset. Cheers!

**Comment by gkalpak** [comment]
I updated the documentation (per https://github.com/angular/angular.js/pull/11719#issuecomment-150535425), made the change described in https://github.com/angular/angular.js/pull/11719#discussion_r43343003, fixed some typos in the error message and merged !

Well done everyone :smiley:
